go/ct/gen/interval_solver.go

```diff
@@ -0,0 +1,188 @@
+		return &IntervalSolver[T]{}
+	}
+	return &IntervalSolver[T]{
+		intervals: []interval[T]{initialInterval},
+	}
+}
+
+func (s *IntervalSolver[T]) Exclude(min, max T) {
+	if max < min {
+		return
+	}
+	res := []interval[T]{}
+	for _, current := range s.intervals {
+		newLeft := interval[T]{current.low, min - 1}
+		newRight := interval[T]{max + 1, current.high}
+
+		switch current.getRelationTo(&interval[T]{min, max}) {
+		case isEqual:
+			continue // we remove this interval
+		case isSubset:
+			res = append(res, newLeft, newRight)
+		case isSuperset:
+			continue // the current interval is dropped
+		case intersectsWithStart:
+			res = append(res, newRight)
+		case intersectsWithEnd:
+			res = append(res, newLeft)
+		default:
+			res = append(res, current)
+		}
+
+	}
+	s.intervals = res
+}
+
+func (s *IntervalSolver[T]) Contains(value T) bool {
+	for _, interval := range s.intervals {
+		if interval.contains(value) {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *IntervalSolver[T]) AddEqualityConstraint(value T) {
+	s.AddLowerBoundary(value)
+	s.AddUpperBoundary(value)
+}
+
+func (s *IntervalSolver[T]) AddLowerBoundary(value T) {
+	if len(s.intervals) == 0 {
+		return
+	}
+	min := s.intervals[0].low
+	s.Exclude(min, value-1)
+}
+
+func (s *IntervalSolver[T]) AddUpperBoundary(value T) {
+	if len(s.intervals) == 0 {
+		return
+	}
+	max := s.intervals[len(s.intervals)-1].high
+	s.Exclude(value+1, max)
+}
+
+func (s *IntervalSolver[T]) IsSatisfiable() bool {
+	return len(s.intervals) > 0
+}
+
+func (s *IntervalSolver[T]) Generate(rnd *rand.Rand) (T, error) {
+	if s.intervals == nil || len(s.intervals) == 0 {
+		return 0, ErrUnsatisfiable
+	}
+
+	domainSize := uint64(0)
+	for _, interval := range s.intervals {
+		if interval.isEmpty() {
+			return 0, ErrUnsatisfiable
+		}
+		domainSize += uint64(interval.high - interval.low + 1)
+	}
+
+	if domainSize == 0 { // the domain is the full uint64 domain
+		return T(rnd.Uint64()), nil
+	}
+
+	sample := rnd.Uint64n(domainSize)
+	for _, interval := range s.intervals {
+		if uint64(interval.high-interval.low+1) > sample {
+			return interval.low + T(sample), nil
+		}
+		sample -= uint64(interval.high - interval.low + 1)
+	}
+	return 0, fmt.Errorf("internal error")
+}
+
+func (s *IntervalSolver[T]) String() string {
+	if len(s.intervals) == 0 {
+		return "false"
+	}
+	clauses := []string{}
+	for _, interval := range s.intervals {
+		clauses = append(clauses, fmt.Sprintf("[%d..%d]", interval.low, interval.high))
+	}
+	return fmt.Sprintf("X ∈ %s", strings.Join(clauses, " ∪ "))
+}
+
+func (s *IntervalSolver[T]) Clone() *IntervalSolver[T] {
+	return &IntervalSolver[T]{intervals: append([]interval[T]{}, s.intervals...)}
+}
+
+func (s *IntervalSolver[T]) Equals(other *IntervalSolver[T]) bool {
+	if len(s.intervals) != len(other.intervals) {
+		return false
+	}
+	for i, interval := range s.intervals {
+		if interval != other.intervals[i] {
+			return false
+		}
+	}
+	return true
```

Comment:
This function is only true if the following invariants are guaranteed:
- the intervals are sorted
- there are no overlaps; in particular, there are no empty intervals
The last one should be the case, but in the print function you are checking for this empty intervals. So this should be ensured.